### PR TITLE
enhancement(file source): Enable skipping first N lines

### LINKED ISSUE #2175: Support skipping first N lines in `file` source
It would be useful to have an option to support skipping first N lines of the file in `file` source. The primary use case is to skip header rows for structured file formats such as TSV.

The feature was requested in the chat:
> Is it possible, when using sources = file, to skip the first N lines from each new file since they have a description of fields?

### PR DESCRIPTION
**Pull request by ktff**
Closes #2175

### Open questions
- [ ] This PR propagates `file_id` which is used to identify a file. It's need for this enhancement to function properly. But a similar enhancement `multiline` uses file path as an identifier which should break under certain file rotation strategies.

### CODE REVIEW

**Review by binarylogic — APPROVED** [review]
Looks great to me! @lukesteensen the code review pleasure is all yours 😄.

**Review by lukesteensen — CHANGES_REQUESTED** [review]
This will skip the first N lines per file that are returned by the file server on each run of Vector, no matter where those lines are located in the file.

To actually skip the first N lines in each physical file I think we'll need something much more invasive and stateful, which is why I was hesitant to prioritize this feature in the first place. Feel free to take a stab at it, but I don't want to invest more time and effort into this feature than it's worth.

**Comment by binarylogic** [comment]
@ktff if you find that the remaining work outweighs the benefits, let's note that on the issue and close it.

**Comment by ktff** [comment]
@lukesteensen then we could only enable this feature if `start_at_beginning` is enabled, this should be enough for the use case in the issue, unless there is some other problem you see?

File: .meta/sources/file.toml.erb [hunk 1]
```diff
@@ -240,6 +240,14 @@ Instead of balancing read capacity fairly across all watched files, prioritize \
 draining the oldest files before moving on to read data from younger files.\
 """
 
+[sources.file.options.skip_first_lines]
+type = "uint"
+common = false
+default = 0
+description = """\
+Number of first lines per file that should be skipped.\
+"""
```
Comment (binarylogic) [inline .meta/sources/file.toml.erb:249]:
```suggestion
"""
relevant_when = {start_at_beginning = true}
```

We should add the above if this is only relevant when `start_at_beginning` is true.

**Comment by ktff** [comment]
@lukesteensen  Lines will now be skipped only if `start_at_beginning` is enabled.

**Comment by lukesteensen** [comment]
I'm still a little nervous about this given that `start_at_beginning` sadly does not guarantee we'll actually start at the beginning of a file (table from #1020): 

Old | Startup | Checkpoint | start_at_beginning | Result
-- | -- | -- | -- | --
Y | - | - | - | End
N | Y | Y | Y | Beginning
N | Y | Y | N | Checkpoint
N | Y | N | Y | Beginning
N | Y | N | N | Beginning
N | N | Y | Y | Checkpoint
N | N | Y | N | Checkpoint
N | N | N | Y | Beginning
N | N | N | N | Beginning

**Comment by ktff** [comment]
> I'm still a little nervous about this

For a good reason, true, this won't work. There are some checks that we could do, but these seam too complex for such a feature so I'll be closing this. At best, once #1020 is done we can come back to this.

**Pull request closed without merging**